### ISSUE TITLE
'run' with continous_read reproduces lines from mixed STDOUT+STDERR output in wrong order

### ISSUE THREAD

**Issue by lumiera** [RexOps/Rex#1171]
Testcase:

Executable file /tmp/outerr.pl on remote host:

```
#!/usr/bin/env perl

use strict;
use warnings;

STDOUT->say('01 stdout');
sleep 1;
STDERR->say('02 stderr');
sleep 1;
STDOUT->say('03 stdout');
```

Rexfile on local host:
```
use strict;
use warnings;
use Rex -feature => [qw/1.3/];

user 'root';
task TTT => sub {
    run '/tmp/outerr.pl', continuous_read => sub { say $_[0] };
};
```

Results of execution:
```
% rex -H remote.host TTT                                                                         
[2017-09-21 11:08:19] INFO - Running task TTT on remote.host
02 stderr
01 stdout
03 stdout
[2017-09-21 11:08:21] INFO - All tasks successful on all hosts
```

Notice that the order of output lines is not correct. Adding 2>&1 to command does not solve this.

**Comment by ferki** [comment]
@lumiera: thanks for the report and your patience!

## TL;DR

It seems related to standard `libc` I/O buffering to me, and I can't think of any way to overcome this from inside Rex. Therefore I decided to close this issue as not planned.

### In case I'm wrong

If any readers are aware of a way to control the output buffering of other programs, please open a [discussion about the idea](https://github.com/RexOps/Rex/discussions/new?category=ideas), ideally with demonstration, and refer to this issue.

## Longer version

The described behavior is not specific to Rex, but related to I/O buffering of `libc`, and can be reproduced by asking rex to execute other programs with a mixed output to both STDOUT and STDERR, like:

```perl
task TTT => sub {
    # list an existing and a non-existing directory with a single command
    # to write to both STDOUT and STDERR
    run 'ls /tmp /tmp2', continuous_read => sub { say $_[0] };
};
```

According to [`man 3 setbuf`](https://linux.die.net/man/3/setbuf):

> The three types of buffering available are unbuffered, block buffered, and line buffered. When an output stream is unbuffered, information appears on the destination file or terminal as soon as written; [...] when it is line buffered characters are saved up until a newline is output or input is read from any stream attached to a terminal device [...]
>
> [...] If a stream refers to a terminal (as _stdout_ normally does) it is line buffered. The standard error stream _stderr_ is always unbuffered by default.

In the `/tmp/outerr.pl` example, it can be fixed by disabling any buffering (by enabling autoflushing) of any output via the [`$OUTPUT_AUTOFLUSH`](https://perldoc.perl.org/perlvar#$OUTPUT_AUTOFLUSH) or [`$|`](https://perldoc.perl.org/perlvar#$%7C) variable above the `print` calls. Like:

```perl
$| = 1;

# equivalent more readable version below
use English;
$OUTPUT_AUTOFLUSH = 1;
```

The program executed by Rex makes its own decisions about its own output buffering. Since Rex doesn't have control over the source code of the executed program, it is ultimately the other program's responsibility to manage its output in a way that is expected by its users, and/or to provide options for its users to control its behavior.
